overlord/ifacestate/hotplug.go

```diff
@@ -84,11 +106,275 @@ func defaultDeviceKey(devinfo *hotplug.HotplugDeviceInfo, keyVersion int) (strin
 }
 
 // HotplugDeviceAdded gets called when a device is added to the system.
-func (m *InterfaceManager) HotplugDeviceAdded(devinfo *hotplug.HotplugDeviceInfo) {
+func (m *InterfaceManager) hotplugDeviceAdded(devinfo *hotplug.HotplugDeviceInfo) {
+	st := m.state
+	st.Lock()
+	defer st.Unlock()
+
+	coreSnapInfo, err := snapstate.CoreInfo(st)
+	if err != nil {
+		logger.Noticef("core snap not available, hotplug events ignored")
+		return
+	}
+
+	hotplugIfaces := m.repo.AllHotplugInterfaces()
+	defaultKey, err := defaultDeviceKey(devinfo, deviceKeyVersion)
+	if err != nil {
+		logger.Noticef(err.Error())
+	}
+
+	hotplugFeature, err := m.hotplugEnabled()
+	if err != nil {
+		logger.Noticef(err.Error())
+		return
+	}
+
+	// iterate over all hotplug interfaces
+	for _, iface := range hotplugIfaces {
+		hotplugHandler := iface.(hotplug.Definer)
+
+		// determine device key for the interface; note that interface might provide own device keys.
+		key, err := deviceKey(defaultKey, devinfo, iface)
+		if err != nil {
+			logger.Noticef(err.Error())
+			continue
+		}
+
+		spec := hotplug.NewSpecification()
+		if hotplugHandler.HotplugDeviceDetected(devinfo, spec) != nil {
+			logger.Noticef("Failed to process hotplug event by the rule of interface %q: %s", iface.Name(), err)
+			continue
+		}
+		slotSpec := spec.Slot()
+		if slotSpec == nil {
+			continue
+		}
+
+		if key == "" {
+			logger.Debugf("No valid hotplug key provided by interface %s, device %q ignored", iface.Name(), devinfo.DeviceName())
+			continue
+		}
+
+		if slotSpec.Label == "" {
+			si := interfaces.StaticInfoOf(iface)
+			slotSpec.Label = si.Summary
+		}
+
+		if !hotplugFeature {
```

Comment:
Could this be done earlier in the loop?

Reply:
Yes, it could, in fact we could avoid entire loop if not enabled, but the initial idea was to do as much as possible with feature disabled (i.e. everything that's not affecting state) to spot potential problems.

Reply:
In fact I tried that, moving this before the loop makes it hard to provide any meaningful message and not flood the logs: hotplugDeviceAdded is called a lot as udev reports various devices; only inside the loop we know if given device is relevant for us at all (if we have any interface that wants to create a hotplug slot for it).